dnssec-tools/validator/doc/getquery.1
.\" Automatically generated by Pod::Man v1.37, Pod::Parser v1.35
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sh \" Subsection heading
.br
.if t .Sp
.ne 5
.PP
\fB\\$1\fR
.PP
..
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  | will give a
.\" real vertical bar.  \*(C+ will give a nicer C++.  Capital omega is used to
.\" do unbreakable dashes and therefore won't be available.  \*(C` and \*(C'
.\" expand to `' in nroff, nothing in troff, for use with C<>.
.tr \(*W-|\(bv\*(Tr
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.Sh), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.if \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.\"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.hy 0
.if n .na
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "GETQUERY 1"
.TH GETQUERY 1 "2011-06-28" "perl v5.8.9" "User Commands"
.SH "NAME"
getquery \- command\-line test program for the \fIval_res_query()\fR function 
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&   getquery [options] name
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
This utility is a command-line wrapper around the \fI\fIval_res_query()\fI\fR
function.  It invokes the \fI\fIval_res_query()\fI\fR function with the given
command-line arguments and displays the returned data as a DNS-style response. 
It also displays the validation status value for the records returned for the query.
.PP
The exit status for this program is 1 if the answer received is trusted, 2 if
the answer received is validated, and \-1 for an error.  The program returns 0 if 
the result was neither trusted nor validated. The trusted and
validated status values are obtained using the \fI\fIval_istrusted()\fI\fR and
\&\fI\fIval_isvalidated()\fI\fR functions from \fI\fIlibval\fI\|(3)\fR.
.SH "OPTIONS"
.IX Header "OPTIONS"
.IP "\-t <type>" 4
.IX Item "-t <type>"
Perform the query for the given \s-1DNS\s0 record type
.IP "\-o, \-\-output=<debug\-level>:<dest\-type>[:<dest\-options>]" 4
.IX Item "-o, --output=<debug-level>:<dest-type>[:<dest-options>]"
<debug\-level> is 1\-7, corresponding to syslog levels ALERT-DEBUG
<dest\-type> is one of file, net, syslog, stderr, stdout
<dest\-options> depends on <dest\-type>
    file:<file\-name>   (opened in append mode)
    net[:<host\-name>:<host\-port>] (127.0.0.1:1053
    syslog[:facility] (0\-23 (default 1 \s-1USER\s0))
.IP "\-h, \-\-help" 4
.IX Item "-h, --help"
Display the help and exit.
.SH "PRE-REQUISITES"
.IX Header "PRE-REQUISITES"
libval
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright 2005\-2011 \s-1SPARTA\s0, Inc.  All rights reserved.
See the \s-1COPYING\s0 file included with the DNSSEC-Tools package for details.
.SH "AUTHORS"
.IX Header "AUTHORS"
Abhijit Hayatnagarkar, Suresh Krishnaswamy, Robert Story, Michael Baer
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fB\f(BIval_res_query\fB\|(3)\fR
.PP
\&\fB\f(BIlibval\fB\|(3)\fR
.PP
http://dnssec\-tools.sourceforge.net


dnssec-tools/NEWS
This file contains a list of major changes per release.  See the
ChangeLog file for a complete set of changes and their details
details.


1.10
 - New Features:
    - New Apps:     (see the validator/apps directory for details)
                    - dnssec-check: check dnssec support from your ISP
                    - dnssec-nodes: graphically displays a DNS
		      hierarchy, color coded by each node's DNSSEC status
                    - dnssec-system-tray: displays pop-up
     		      notifications when a libval-enabled application
		      triggers a DNSSEC error
                    - lookup: a graphical DNS lookup utility that
		      displays the results in a hierarchical tree and
		      color codes the window according to DNSSEC status

    - libval:       - Added support for building on Windows.
                    - added support for falling back to recursion when
                      the caching name server does not appear to 
                      support DNSSEC. This also works as a mechanism
                      to work around poisoned or misbehaving cache. 
                    - Significant improvements to the the asynchronous support. 
    - lsdnssec:     - Improvements to lsdnssec to display different
                      output depending on whether a zone is a
                      stand-alone zone or under control of rollerd.
    - nagios:       - Plugins for the nagios monitoring system which
                      enable monitoring of zone rollerover states.
    - firefox:      - Improved patches that work with the most recent firefox

 Plus many more minor features and bug fixes

1.9
 - New Features:
    - lsdnssec:     - Added a new flag (-p) to show only zones in a
                      particular rollerd phase.
                    - fixed bugs to align timing output with rollerd.
    - rollerd:      - Added a -logtz flag for logging timezones
                    - fixed bugs related to the -alwayssign flag.
		    - zonesigner's path is taken from the config file.
    - rollctl:      - Added -rollall and -rollzone options.
    - zonesigner:   - Assumes keys need to be generated for new zones
		      (Assumes -genkeys option was given if a keyrec file
		      can't be found.)
                    - Exits with unique exit codes if a failure occurs.
                      ("zonesigner -xc CODE" can lookup a description for it.")
		    - Added the -phase option so rollover options could be
		      more easily specified.
    - lights:       - A simple GUI to check the status of rollover states
    - blinkenlights:- Added hide/show commands for rollrec names and zone
		      names, for split-zone support
    - cleankrf:	    - Fixed deletion of obsolete set keyrecs.
    - GUI commands: - Fixed how the Exit command works so they don't coredump.

    - libsres
      & libval:     - New beta support for issuing asynchronous requests.
                      This can speed up queries by up to 4 times if used.
                      (see example code in validator/apps/validator_selftest.c)
                    - NSEC3, DLV and IPv6 are enabled by default.
                    - improved logging and logging-callback support.
    - drawvalmap    - Can output PNG files now

 - Packaging:
                    - Our download page now allows you to download
                      the C validator libraries independently of the
                      full DNNSEC-Tools tool-suite.

 - Many bugs were also fixed in the 240+ changes.

1.8
 - New Features:
    - zonesigner, rollerd
                    - Made changes so that these tools are more compatible 
                      with recent versions of Bind
                    - The zone_errors configuration parameter allows a zone-
                      specific maximum to be set.  Once exceeded, that zone
                      will be skipped rather than allowing rollover to continue.
    - blinkenlights
                    - Recognizes when rollerd abruptly quits, so error messages
                      aren't spewed interminably.
    - ZonFile::Fast - Fixed parsing of DS records containing spaces and 
                      parsing of mname and rname SOA fields
                    - Added support for parsing KEY records
    - keyrec.pm     - Made changes to properly lock keyrec files before 
                      writing to them.
                    - Begun process of deprecating keyrec_open().
    - mapper:       - added a new option: --node-size for mapping
                      complex zones.
    - dnspktflow:   - added two new options:
                      --layout-style for selecting the layout style to use
                      --node-size for mapping complex zones.
                    - Add new (default) option to cluster
                      authoritative nodes together to help better
                      understand the relationships between traffic
                      patterns and authoritative name server/zone arrangement.
    - libval:       - Now distributed with the Root TA.
                    - Added stricter checks for openssl SHA-256 support in 
                      configure. 
                    - Added several improvements that allow the validator to 
                      lookup information within provably insecure zones that 
                      do not handle EDNS0 requests nicely. This includes 
                      adding support for turning off EDNS0 when traversing a 
                      name hierarchy that leads to a provably insecure zone, 
                      EDNS0 fallback support, and additional checks to check 
                      the sanity of response data.
                    - Fixed certain bugs in CNAME handling and in the 
                      validation of proofs accompanying wildcard responses, 
                      referrals and alias chains.
                    - Fixed support for RSADSA and RSASHA-512 signature 
                      validation.

dnssec-tools/validator/apps/getquery.c
/*
 * Copyright 2005-2011 SPARTA, Inc.  All rights reserved.
 * See the COPYING file distributed with this software for details.
 *
 * A command-line tool for testing the val_res_query() function.
 */

#include "validator-config.h"
#include <validator/validator.h>
#include <validator/resolver.h>

#ifdef HAVE_GETOPT_H
#include <getopt.h>
#endif

#define	NAME	"getquery"
#define	VERS	"version: 1.0"
#define	DTVERS	"DNSSEC-Tools Version: 1.8"

#ifdef HAVE_GETOPT_LONG
// Program options
static struct option prog_options[] = {
    {"help", 0, 0, 'h'},
    {"type", 0, 0, 't'},
    {"output", 0, 0, 'o'},
    {"Version", 0, 0, 'V'},
    {0, 0, 0, 0}
};
#endif

void
usage(char *progname)
{
    fprintf(stderr,
            "Usage: %s [options] hostname\n",
            progname);
    fprintf(stderr, "Options:\n");
    fprintf(stderr,
            "\t-h, --help          display usage and exit\n");
    fprintf(stderr,
            "\t-t, --type=<type>   record type. Defaults to A record.\n");
    fprintf(stderr,
            "\t-o, --output=<debug-level>:<dest-type>[:<dest-options>]\n"
            "\t          <debug-level> is 1-7, corresponding to syslog levels\n"
            "\t          <dest-type> is one of file, net, syslog, stderr, stdout\n"
            "\t          <dest-options> depends on <dest-type>\n"
            "\t              file:<file-name>   (opened in append mode)\n" 
            "\t              net[:<host-name>:<host-port>] (127.0.0.1:1053\n" 
            "\t              syslog[:facility] (0-23 (default 1 USER))\n" );
    fprintf(stderr,
            "\t-V, --Version       display version and exit\n");
}

void
version(void)
{
    fprintf(stderr, "%s: %s\n", NAME,VERS);
    fprintf(stderr, "%s\n", DTVERS);
}

int
main(int argc, char *argv[])
{
    char           *node = NULL;
    val_log_t      *logp;
    u_int16_t      type_h = ns_t_a;
    int success = 0;
    int ret;
    val_status_t status;
    u_char buf[1024];

    while (1) {
        int             c;
#ifdef HAVE_GETOPT_LONG
        int             opt_index = 0;
#ifdef HAVE_GETOPT_LONG_ONLY
        c = getopt_long_only(argc, argv, "ho:t:V",
                             prog_options, &opt_index);
#else
        c = getopt_long(argc, argv, "ho:t:V", prog_options, &opt_index);
#endif
#else                           /* only have getopt */
        c = getopt(argc, argv, "ho:t:V");
#endif

        if (c == -1) {
            break;
        }

        switch (c) {
        case 'h':
            usage(argv[0]);
            return -1;
        case 'o':
            logp = val_log_add_optarg(optarg, 1);
            if (NULL == logp) { /* err msg already logged */
                usage(argv[0]);
                return -1;
            }
            break;

        case 't':
            type_h = res_nametotype(optarg, &success);
            if (!success) {
                fprintf(stderr, "Unrecognized type %s\n", optarg);
                usage (argv[0]);
                return -1;
            }
            break;
        case 'V':
            version();
            return 0;

        default:
            fprintf(stderr, "Invalid option %s\n", argv[optind - 1]);
            usage(argv[0]);
            return -1;
        }
    }

    if (optind < argc) {
        node = argv[optind++];
    } else {
        fprintf(stderr, "Error: node name not specified\n");
        usage(argv[0]);
        return -1;
    }

    //ret = val_res_query(NULL, "mail.marzot.net", 1, 15, buf, 1024, &status);
    //ret = val_res_query(NULL, "nosig-A.test.dnssec-tools.org.", 1, 1, buf, 1024, &status);
    //ret = val_res_query(NULL, "good-A.test.dnssec-tools.org", 1, 28, buf, 1024, &status);
    //ret = val_res_query(NULL, "vb.futz.org", 1, 1, buf, 1024, &status);
    
    ret = val_res_query(NULL, node, ns_c_in, type_h, buf, 1024, &status);
    printf("Return value %d\n", ret);
    printf("herrno value %d\n", h_errno);
    printf("Validation Status %s\n", p_val_status(status));
    if (ret > 0) {
        print_response(buf, ret);
    }

    if (val_isvalidated(status))
        return 2;
    else if (val_istrusted(status))
        return 1;
    return 0;
}
